Batch add index block cluster state updates

joegallo opened the pull request:
Related to https://github.com/elastic/elasticsearch/issues/81627 and https://github.com/elastic/elasticsearch/pull/83432, followup to https://github.com/elastic/elasticsearch/pull/83760 and https://github.com/elastic/elasticsearch/pull/84259.

joegallo committed:
Switch to listener.delegateFailure(...)

joegallo committed:
Whitespace

joegallo committed:
Extract an executor for adding blocks

joegallo committed:
Do the same for finalizing blocks

elasticmachine commented:
Pinging @elastic/es-data-management (Team:Data Management)

joegallo committed:
Update docs/changelog/84374.yaml

elasticsearchmachine commented:
Hi @joegallo, I've created a changelog YAML for you.

joegallo committed:
Merge branch 'master' into batch-add-blocks

joegallo commented:
@elasticmachine run elasticsearch-ci/part-2

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java:
```diff
@@ -480,87 +483,125 @@ public void addIndexBlock(AddIndexBlockClusterStateUpdateRequest request, Action
+                    final var task = taskContext.getTask();
+                    final Tuple<ClusterState, Map<Index, ClusterBlock>> blockResult = addIndexBlock(
+                        task.request.indices(),
                         currentState,
```
Oops

```suggestion
                        state,
```

Otherwise we're not accumulating updates. Could we have a test to catch this?

joegallo replied:
OOF.

Great catch, how embarrassing. I'll get a test in place and fix it. Thanks for catching this!

joegallo replied:
How does f90215b79778fdf63c365eb38beecfb5f7a657a2 look to you, @DaveCTurner?

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java:
```diff
@@ -480,87 +483,125 @@ public void addIndexBlock(AddIndexBlockClusterStateUpdateRequest request, Action
+        }
+    }
+
+    private class FinalizeBlocksExecutor implements ClusterStateTaskExecutor<FinalizeBlocksTask> {
```
Nit: 

```suggestion
    private static class FinalizeBlocksExecutor implements ClusterStateTaskExecutor<FinalizeBlocksTask> {
```

joegallo replied:
035e71a8b7eb50b276ad62546d7928c9507e56f0, sure, this one doesn't have to not be static (unlike the other Executors in this file, where they do happen to rely on instance fields of the enclosing class).

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java:
```diff
@@ -480,87 +483,125 @@ public void addIndexBlock(AddIndexBlockClusterStateUpdateRequest request, Action
+            for (final var taskContext : taskContexts) {
+                try {
+                    final var task = taskContext.getTask();
+                    final Tuple<ClusterState, List<AddBlockResult>> closingResult = finalizeBlock(
```
Naming nit:

```suggestion
                    final Tuple<ClusterState, List<AddBlockResult>> finalizeResult = finalizeBlock(
```

joegallo replied:
389de77faeec5ec54c1eaace61f69e5bfec918b1, oops!

DaveCTurner reviewed (commented):
Looks good except one slip that I think indicates a lack of test coverage (and a couple of tiny nits)

joegallo committed:
Rename this variable, oops

joegallo committed:
Make this class static

The rest of the various Executors here are not static, because they
use the clusterService or some other field of the enclosing
MetadataIndexStateService, but this one doesn't happen to do that, so
there's no harm in making it static.

joegallo committed:
Merge branch 'master' into batch-add-blocks

joegallo committed:
Merge branch 'master' into batch-add-blocks

joegallo committed:
Add a failing test

joegallo committed:
Make it pass

joegallo committed:
Merge branch 'master' into batch-add-blocks

DaveCTurner commented on server/src/test/java/org/elasticsearch/cluster/metadata/MetadataIndexStateServiceBatchingTests.java:
```diff
@@ -0,0 +1,207 @@
+        final var future2 = client().admin().indices().prepareOpen("test-2", "test-3").execute();
+
+        // check the queue for the open-indices tasks
+        assertBusy(() -> assertThat(masterService.pendingTasks(), hasSize(3))); // two plus the blocking task itself
```
This assertion should hold immediately, no need for an `assertBusy()` I think.

joegallo replied:
I didn't think it would necessarily hold immediately, what with the `.execute()` methods returning `ActionFuture`s -- are there any guarantees how of how much processing they'll do before they return?

DaveCTurner commented on server/src/test/java/org/elasticsearch/cluster/metadata/MetadataIndexStateServiceBatchingTests.java:
```diff
@@ -0,0 +1,207 @@
+
+        // assert that the requests were acknowledged
+        assertAcked(future1.get());
+        assertAcked(future2.get());
```
Could we start by closing (some random subset of) the indices and then assert that they're all open again here? That seems important.

joegallo replied:
Oh, lol, they were all supposed to be closed indices.

joegallo replied:
7d8cbb9ecf2 to make them all closed indices as I had originally intended, but another commit incoming that'll make only a random subset of them be closed.

joegallo replied:
2762ad600c2 to check that they're really open, though.

joegallo replied:
de904285da09766f41ddb8ab7389678a1b0c9d5a for the randomization

DaveCTurner commented on server/src/test/java/org/elasticsearch/cluster/metadata/MetadataIndexStateServiceBatchingTests.java:
```diff
@@ -0,0 +1,207 @@
+        final var future2 = client().admin().indices().prepareClose("test-2", "test-3").execute();
+
+        // check the queue for the first close tasks (the add-block-index-to-close tasks)
+        assertBusy(() -> assertThat(masterService.pendingTasks(), hasSize(3))); // two plus the blocking task itself
```
Similarly, this should hold immediately.

joegallo replied:
4137dcaf9080b3df9517a1f348f91122ab9d625c

DaveCTurner commented on server/src/test/java/org/elasticsearch/cluster/metadata/MetadataIndexStateServiceBatchingTests.java:
```diff
@@ -0,0 +1,207 @@
+        final var resp2 = future2.get();
+        assertAcked(resp2);
+        assertThat(resp2.getIndices(), hasSize(2));
+        assertThat(resp2.getIndices().stream().map(r -> r.getIndex().getName()).toList(), containsInAnyOrder("test-2", "test-3"));
```
Similarly, we should be able to assert that the indices were genuinely closed here.

joegallo replied:
👍, 2762ad600c2

DaveCTurner commented on server/src/test/java/org/elasticsearch/cluster/metadata/MetadataIndexStateServiceBatchingTests.java:
```diff
@@ -0,0 +1,207 @@
+        final var future2 = client().admin().indices().prepareAddBlock(APIBlock.WRITE, "test-2", "test-3").execute();
+
+        // check the queue for the first add-block tasks (the add-index-block tasks)
+        assertBusy(() -> assertThat(masterService.pendingTasks(), hasSize(3))); // two plus the blocking task itself
```
Similarly, this shouldn't need the `assertBusy()`.

DaveCTurner commented on server/src/test/java/org/elasticsearch/cluster/metadata/MetadataIndexStateServiceBatchingTests.java:
```diff
@@ -0,0 +1,207 @@
+        final var resp2 = future2.get();
+        assertAcked(resp2);
+        assertThat(resp2.getIndices(), hasSize(2));
+        assertThat(resp2.getIndices().stream().map(r -> r.getIndex().getName()).toList(), containsInAnyOrder("test-2", "test-3"));
```
Similarly, could we assert that the blocks were added?

joegallo replied:
👍, 2762ad600c2

DaveCTurner reviewed (commented):
That's what I had in mind, yes. Added some small suggestions. Also could we add a cluster state listener in each case so we can assert how many cluster states get published?

DaveCTurner commented on server/src/test/java/org/elasticsearch/cluster/metadata/MetadataIndexStateServiceBatchingTests.java:
```diff
@@ -0,0 +1,207 @@
+        assertThat(resp2.getIndices().stream().map(r -> r.getIndex().getName()).toList(), containsInAnyOrder("test-2", "test-3"));
+    }
+
+    private static Callable<Void> blockMasterService(MasterService masterService) {
```
Why `Callable<Void>` and not just `CheckedRunnable<Exception>`?

joegallo replied:
Why, because I didn't know that existed of course. 😆

joegallo replied:
bb3839552d7c254e1b7b1cb395f6d49bb9cc0f72

joegallo committed:
Drop some unnecessary assertBusy blocks

joegallo committed:
Switch to CheckedRunnable

joegallo committed:
Test opens against *closed* indices

joegallo committed:
Check post conditions

joegallo committed:
Merge branch 'master' into batch-add-blocks

DaveCTurner commented on server/src/test/java/org/elasticsearch/cluster/metadata/MetadataIndexStateServiceBatchingTests.java:
```diff
@@ -0,0 +1,232 @@
+
+        // and then close them (and wait for the close to actually run)
+        assertAcked(client().admin().indices().prepareClose("test-1", "test-2", "test-3"));
+        assertBusy(() -> assertThat(masterService.pendingTasks(), hasSize(0)));
```
Hmm. We apparently need to wait for the `shard-started` tasks to complete, but there might be an intermediate state with zero pending tasks during that process too I think. `ensureGreen` would seem safer? (Also you don't need the previous `ensureGreen`, you can close the index straight away).

joegallo replied:
Sure thing, 69d9f724524cd99876dc7ce73f9fae0fdeedd5c8

joegallo committed:
Move the ensureGreen, drop the assertBusy

joegallo committed:
Randomly close some indices before testing opens

joegallo commented:
> Also could we add a cluster state listener in each case so we can assert how many cluster states get published?

It looks to me like shard-started tasks end up making this a little harder than it sounds at first glance -- it looks to me like they're not consistently batching in a way that makes this amenable to the super simple "just count all changed states".

Let's catch up on Monday and see if there's something we can do to be more clever and get around that.

DaveCTurner commented:
Oh right yes that's not going to work. How about assertions that there are never any partially-completed states?

```diff
diff --git a/server/src/test/java/org/elasticsearch/cluster/metadata/MetadataIndexStateServiceBatchingTests.java b/server/src/test/java/org/elasticsearch/cluster/metadata/MetadataIndexStateServiceBatchingTests.java
index b52eae61e5e..63c7188d7a3 100644
--- a/server/src/test/java/org/elasticsearch/cluster/metadata/MetadataIndexStateServiceBatchingTests.java
+++ b/server/src/test/java/org/elasticsearch/cluster/metadata/MetadataIndexStateServiceBatchingTests.java
@@ -10,6 +10,7 @@ package org.elasticsearch.cluster.metadata;
 
 import org.elasticsearch.action.ActionListener;
 import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.ClusterStateListener;
 import org.elasticsearch.cluster.ClusterStateTaskConfig;
 import org.elasticsearch.cluster.ClusterStateTaskListener;
 import org.elasticsearch.cluster.metadata.IndexMetadata.APIBlock;
@@ -30,6 +31,7 @@ import static org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertAcke
 import static org.hamcrest.Matchers.containsInAnyOrder;
 import static org.hamcrest.Matchers.hasSize;
 import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.oneOf;
 
 public class MetadataIndexStateServiceBatchingTests extends ESSingleNodeTestCase {
 
@@ -60,6 +62,9 @@ public class MetadataIndexStateServiceBatchingTests extends ESSingleNodeTestCase
         assertAcked(client().admin().indices().prepareClose(closedIndices));
         ensureGreen("test-1", "test-2", "test-3");
 
+        final var assertingListener = closedIndexCountListener(closedIndices.length);
+        clusterService.addListener(assertingListener);
+
         final var block1 = blockMasterService(masterService);
         block1.run(); // wait for block
 
@@ -81,12 +86,17 @@ public class MetadataIndexStateServiceBatchingTests extends ESSingleNodeTestCase
             final var indexMetadata = clusterService.state().metadata().indices().get(index);
             assertThat(indexMetadata.getState(), is(State.OPEN));
         }
+
+        clusterService.removeListener(assertingListener);
     }
 
     public void testBatchCloseIndices() throws Exception {
         final var clusterService = getInstanceFromNode(ClusterService.class);
         final var masterService = clusterService.getMasterService();
 
+        final var assertingListener = closedIndexCountListener(3);
+        clusterService.addListener(assertingListener);
+
         // create some indices
         createIndex("test-1", client().admin().indices().prepareCreate("test-1"));
         createIndex("test-2", client().admin().indices().prepareCreate("test-2"));
@@ -130,12 +140,17 @@ public class MetadataIndexStateServiceBatchingTests extends ESSingleNodeTestCase
             final var indexMetadata = clusterService.state().metadata().indices().get(index);
             assertThat(indexMetadata.getState(), is(State.CLOSE));
         }
+
+        clusterService.removeListener(assertingListener);
     }
 
     public void testBatchBlockIndices() throws Exception {
         final var clusterService = getInstanceFromNode(ClusterService.class);
         final var masterService = clusterService.getMasterService();
 
+        final var assertingListener = blockedIndexCountListener();
+        clusterService.addListener(assertingListener);
+
         // create some indices
         createIndex("test-1", client().admin().indices().prepareCreate("test-1"));
         createIndex("test-2", client().admin().indices().prepareCreate("test-2"));
@@ -179,6 +194,8 @@ public class MetadataIndexStateServiceBatchingTests extends ESSingleNodeTestCase
             final var indexMetadata = clusterService.state().metadata().indices().get(index);
             assertThat(INDEX_BLOCKS_WRITE_SETTING.get(indexMetadata.getSettings()), is(true));
         }
+
+        clusterService.removeListener(assertingListener);
     }
 
     private static CheckedRunnable<Exception> blockMasterService(MasterService masterService) {
@@ -200,6 +217,17 @@ public class MetadataIndexStateServiceBatchingTests extends ESSingleNodeTestCase
         return () -> executionBarrier.await(10, TimeUnit.SECONDS);
     }
 
+    private static ClusterStateListener closedIndexCountListener(int closedIndices) {
+        return event -> assertThat(event.state().metadata().getConcreteAllClosedIndices().length, oneOf(0, closedIndices));
+    }
+
+    private static ClusterStateListener blockedIndexCountListener() {
+        return event -> assertThat(
+            event.state().metadata().stream().filter(indexMetadata -> INDEX_BLOCKS_WRITE_SETTING.get(indexMetadata.getSettings())).count(),
+            oneOf(0L, 3L)
+        );
+    }
+
     /**
      * Listener that asserts it does not fail.
      */
```

joegallo committed:
Add listeners to assert invariants

That is, this is checking that we're *actually* batching these
operations

joegallo committed:
Merge branch 'master' into batch-add-blocks

joegallo commented:
👍, @DaveCTurner I like that approach. Added as a44f7451fe875a44723974c11f1649d815485b12.

DaveCTurner reviewed (approved):
LGTM
